### DISCUSSION TITLE
Focusing a Text Input

### DISCUSSION THREAD

**Discussion by esegebart** [react-native-tvos/react-native-tvos#338] [General Discussions]
I have a login screen on my TV app with a text input wrapped in a TouchableHighlight to access the onFocus, onPress, and onBlur. I can't seem to figure out how to get these text inputs to show that they are focused. I can click on them and the keyboard pops up. I just cant see where the arrow keys on my Firestick remote are navigating to. If I click twice, I make it down to the password. If I click three more times, I get to the login button and can sign in. Any ideas on how to show the focus on text inputs? 

Here is an example of how I am trying to do it. It works - it just doesn't show me the focus on the inputs.

```
const onFocus = (inputRef) => {
    setFocused(true);
    inputRef.current.focus();
    console.log('focused?', focused);
    setEmail(email);
  };

  const onBlur = (inputRef) => {
    setFocused(false);
    console.log('focused?', focused);
  };

<TouchableHighlight
    onPress={() => onFocus(emailRef)}
    onBlur={onBlur}
    hasTVPreferredFocus
  >
    <TextInput 
      ref={emailRef}
      type="email"
      style={focused ? styles.emailFocused : styles.textInput}
      value={email}
      onChangeText={(v) => setEmail(v.trim())}
      placeholder="EMAIL ADDRESS"
      placeholderTextColor={"#bcbcbc"}
      color={"black"}
    />
</TouchableHighlight>
```

**Comment by nriccar** [comment]
What do you mean by "show the focus on the input"? I see you are setting a style based on the focused state, isn't it working?

**Reply by esegebart** [reply]
I cannot seem to get it to work with the styling when I navigate with the remote. If you think styling would be the way to go, I will do a bit more tinkering and see if I can get it.

**Reply by esegebart** [reply]
Actually - I am running the wrong project. Let me try the right one and get back to you.

**Reply by esegebart** [reply]
@nriccar Thank you for responding.. it seems I had my correct project open in VS Code and the wrong project open in Android Studio 😅 Navigation looks fantastic and works like a charm. Thank you!

**Reply by nriccar** [reply]
That's it! Happy to help :)

**Reply by nriccar** [reply]
Just a friendly suggestion, play with this props for Touchables, you can accomplish some crazy things:
```javascript
     * enabled: If true, parallax effects are enabled.  Defaults to true.
     * shiftDistanceX: Defaults to 2.0.
     * shiftDistanceY: Defaults to 2.0.
     * tiltAngle: Defaults to 0.05.
     * magnification: Defaults to 1.0.
     * pressMagnification: Defaults to 1.0.
     * pressDuration: Defaults to 0.3.
     * pressDelay: Defaults to 0.0.
```

**Reply by esegebart** [reply]
Ooooo wow! I will definitely mess around with these! Thank you!

**Reply by nriccar** [reply]
they go under the parallaxProperties on the touchable. Happy to help 😊

**Reply by esegebart** [reply]
@nriccar I was wondering if you could help me. I am trying to navigate to all of my touchables on my page and the focus is jumping up to another section when I navigate right every other press. Here is how I have it set up: 

So while testing my app, I have a header component that has two touchables that I can navigate from. When I push down from the first touchable, it doesn't go anywhere. When I press down from the second one, it goes to the next line that has a carousel with videos.

The carousel component that has several videos in it. When I try to navigate right to the next video, it skips over it and focuses on every other one, after two presses, it jumps up to the line of videos above it.. I can't seem to get the focus to navigate to each video individually and stay on the same line of videos I am browsing through. 

Do you know where I can find how to focus each thing properly?

**Reply by nriccar** [reply]
Of course I can! 

Touchables directions can be tricky.. but once you got how they work, or at least a general idea of it, it starts to click..
First of all, as a "good practice" try to have the touchables right next to each other, and, if you can, try to set them the same size (you can accomplish this if you set the with or height to the touchable, and the style you want to a children View component).

If after that the direction is still not working, I would recommend setting the nextDirection props:

**Reply by nriccar** [reply]
https://github.com/react-native-tvos/react-native-tvos/blob/efaa116881b47c75bc77a7538579ad8ab09e6e33/packages/rn-tester/js/examples/DirectionalNextFocus/DirectionalNextFocusExample.js 

Here's an example.. this way you can always override the behaviour of the next focus in the direction you want :)

**Reply by esegebart** [reply]
@nriccar Amazing! Thank you so much. So the two touchables at the top of my page ARE the same size, but I cannot have the other ones the same size because they are for the videos and are much larger. But they ARE right next to each other! And the carousel with my videos are right next to each other also! 

So I wonder if it would be better to have a menu button that contains all of my search, bookmark, etc so all the touchables are together there and the same size instead of having two touchables in my header. 

I will look into the nextDirection set up and see what I can do! Thank you very much for your help. I will let you know if I have anymore questions!

**Reply by esegebart** [reply]
@nriccar Hello again, I am still stuck on the directional focus and how to override it. I may not have enough experience to kind of understand what's going on.. specifically in this part: 

```
 const tagForDestination = (destination: ?React.Node) =>
    destination ? findNodeHandle(destination) : undefined;
```
Is this how it is finding the next touchable? The problem I am having is I have two touchables in a header and a text input "below" the header. But the down arrow works on the second touchable. It seems that the directions think it is linear. I can navigate right from the first touchable, but not down. And from the second touchable I can press the right arrow and it will go "down" to the text input but can also press down to get to the text input too. Nothing is different between the first and the second one, why would it be I can't also press down  from the first one? 

I have added a picture so you can see the different touchables.

If you can provide any more guidance on how to override the direction of touchables, I would sure appreciate it. 

![image](https://user-images.githubusercontent.com/36378900/193612315-71de78ea-940e-44c5-a2f5-2a71ac673ee5.png)
